core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -418,102 +422,120 @@ public HashDistributionTaskSource(
+            addSuccessCallback(splitBatchFuture, splitBatch -> {
+                getSplitTimeRecorder.accept(start);
 
-            List<TaskDescriptor> result = postprocessTasks(partitionTasks.build());
+                for (Split split : splitBatch.getSplits()) {
+                    int bucket = bucketNodeMap.getBucket(split);
+                    int partition = getPartitionForBucket(bucket);
 
-            finished = true;
-            return result;
+                    if (!bucketNodeMap.isDynamic()) {
+                        HostAddress requiredAddress = bucketNodeMap.getAssignedNode(split).get().getHostAndPort();
+                        Set<HostAddress> existingRequirement = partitionToNodeMap.get(partition);
+                        if (existingRequirement.isEmpty()) {
+                            existingRequirement.add(requiredAddress);
+                        }
+                        else {
+                            checkState(
+                                    existingRequirement.contains(requiredAddress),
+                                    "Unable to satisfy host requirement for partition %s. Existing requirement %s; Current split requirement: %s;",
+                                    partition,
+                                    existingRequirement,
+                                    requiredAddress);
+                            existingRequirement.removeIf(host -> !host.equals(requiredAddress));
+                        }
+                    }
+
+                    if (!split.isRemotelyAccessible()) {
+                        Set<HostAddress> requiredAddresses = ImmutableSet.copyOf(split.getAddresses());
+                        verify(!requiredAddresses.isEmpty(), "split is not remotely accessible but the list of addresses is empty: %s", split);
+                        Set<HostAddress> existingRequirement = partitionToNodeMap.get(partition);
+                        if (existingRequirement.isEmpty()) {
+                            existingRequirement.addAll(requiredAddresses);
+                        }
+                        else {
+                            Set<HostAddress> intersection = Sets.intersection(requiredAddresses, existingRequirement);
+                            checkState(
+                                    !intersection.isEmpty(),
+                                    "Unable to satisfy host requirement for partition %s. Existing requirement %s; Current split requirement: %s;",
+                                    partition,
+                                    existingRequirement,
+                                    requiredAddresses);
+                            partitionToNodeMap.replaceValues(partition, ImmutableSet.copyOf(intersection));
+                        }
+                    }
+
+                    Multimap<PlanNodeId, Split> partitionSplits = partitionToSplitsMap.computeIfAbsent(partition, (p) -> ArrayListMultimap.create());
```

Comment:
`processSplit` is called for multiple split sources. The split loading may happen concurrently from different threads what makes the access to `partitionToSplitsMap` unsafe.

I would suggest to model it slightly differently.

How about having a `ListenableFuture<LoadedSplits> loadSplits(PlanNodeId planNodeId, SplitSource splitSource)`  method that is responsible for loading **all** splits for a given split source and returning an `LoadedSplits` object with the `List<Split> getSplits` and `PlanNodeId getPlanNodeId()` methods.

Then all the futures could be wrapped into `allAsList` and post processed with the `Futures#transform` in a safe manner.